// @stepzen/sdk/src/commands/account.ts
// Copyright (c) 2020,2021,2022, StepZen, Inc.

import * as debug from 'debug'
import fetch from 'node-fetch'

import {ADMIN_ACCOUNT_URL} from '../shared/constants'
import {getRequestHeaders} from '../shared/request'
import {SDKConfiguration, StepZenAccount} from '../shared/types'

export default async (
  account: StepZenAccount,
  sdkConfig: SDKConfiguration,
): Promise<
  {account: string; apikey: string} | {success: false; errors: string[]}
> => {
  const headers = getRequestHeaders(account, sdkConfig)
  const url = `${account.server}${ADMIN_ACCOUNT_URL}`

  debug('stepzen:headers')(headers)

  const response = await fetch(url, {
    headers: headers as any,
    method: 'POST',
  })

  debug('stepzen:response')(response)

  if (response.status >= 400 && response.status < 500) {
    throw new Error(
      'Could not complete the request. Please check your authentication details are correct.',
    )
  }

  if (response.status >= 500) {
    return {
      errors: [response.statusText],
      success: false,
    }
  }

  const text = await response.text()
  let json

  try {
    json = JSON.parse(text)
  } catch {
    return {
      errors: [`An unexpected error occurred.\n\n${text}`],
      success: false,
    }
  }

  return json
}
