// @stepzen/transpiler/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate = exports.transpile = exports.stitch = exports.print = exports.merge = exports.lint = exports.configure = void 0;
var configure_1 = require("./actions/configure");
Object.defineProperty(exports, "configure", { enumerable: true, get: function () { return configure_1.default; } });
var lint_1 = require("./actions/lint");
Object.defineProperty(exports, "lint", { enumerable: true, get: function () { return lint_1.default; } });
var merge_1 = require("./actions/merge");
Object.defineProperty(exports, "merge", { enumerable: true, get: function () { return merge_1.default; } });
var print_1 = require("./actions/print");
Object.defineProperty(exports, "print", { enumerable: true, get: function () { return print_1.default; } });
var stitch_1 = require("./actions/stitch");
Object.defineProperty(exports, "stitch", { enumerable: true, get: function () { return stitch_1.default; } });
var transpile_1 = require("./actions/transpile");
Object.defineProperty(exports, "transpile", { enumerable: true, get: function () { return transpile_1.default; } });
var validate_1 = require("./actions/validate");
Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return validate_1.default; } });
